stl/inc/format

```diff
@@ -2859,101 +2862,107 @@ _NODISCARD auto make_wformat_args(const _Args&... _Vals) {
+        _Parse_format_string(_Fmt, _Handler);
+        return _Buf._Out();
+    }
 }
 
 template <output_iterator<const char&> _OutputIt>
-_OutputIt vformat_to(_OutputIt _Out, const locale& _Loc, const string_view _Fmt,
-    const format_args_t<type_identity_t<_OutputIt>, char> _Args) {
-    _Format_handler<_OutputIt, char, basic_format_context<_OutputIt, char>> _Handler(
-        _STD move(_Out), _Fmt, _Args, _Lazy_locale{_Loc});
-    _Parse_format_string(_Fmt, _Handler);
-    return _Handler._Ctx.out();
+_OutputIt vformat_to(_OutputIt _Out, const locale& _Loc, const string_view _Fmt, const format_args _Args) {
+    if constexpr (is_same_v<_OutputIt, _Fmt_it>) {
+        _Format_handler<char> _Handler(_Out, _Fmt, _Args, _Lazy_locale{_Loc});
+        _Parse_format_string(_Fmt, _Handler);
+        return _Out;
+    } else {
+        _Fmt_iterator_buffer<_OutputIt, char> _Buf(_STD move(_Out));
+        _Format_handler<char> _Handler(_Fmt_it{_Buf}, _Fmt, _Args, _Lazy_locale{_Loc});
+        _Parse_format_string(_Fmt, _Handler);
+        return _Buf._Out();
+    }
 }
 
 template <output_iterator<const wchar_t&> _OutputIt>
-_OutputIt vformat_to(_OutputIt _Out, const locale& _Loc, const wstring_view _Fmt,
-    const format_args_t<type_identity_t<_OutputIt>, wchar_t> _Args) {
-    _Format_handler<_OutputIt, wchar_t, basic_format_context<_OutputIt, wchar_t>> _Handler(
-        _STD move(_Out), _Fmt, _Args, _Lazy_locale{_Loc});
-    _Parse_format_string(_Fmt, _Handler);
-    return _Handler._Ctx.out();
+_OutputIt vformat_to(_OutputIt _Out, const locale& _Loc, const wstring_view _Fmt, const wformat_args _Args) {
+    if constexpr (is_same_v<_OutputIt, _Fmt_wit>) {
+        _Format_handler<wchar_t> _Handler(_Out, _Fmt, _Args, _Lazy_locale{_Loc});
+        _Parse_format_string(_Fmt, _Handler);
+        return _Out;
+    } else {
+        _Fmt_iterator_buffer<_OutputIt, wchar_t> _Buf(_STD move(_Out));
+        _Format_handler<wchar_t> _Handler(_Fmt_wit{_Buf}, _Fmt, _Args, _Lazy_locale{_Loc});
+        _Parse_format_string(_Fmt, _Handler);
+        return _Buf._Out();
+    }
 }
 
 template <output_iterator<const char&> _OutputIt, class... _Types>
 _OutputIt format_to(_OutputIt _Out, const string_view _Fmt, const _Types&... _Args) {
-    _Fmt_iterator_buffer<_OutputIt, char> _Buf(_STD move(_Out));
-    _STD vformat_to(_Fmt_it{_Buf}, _Fmt, _STD make_format_args(_Args...));
-    return _Buf._Out();
+    return _STD vformat_to(_STD move(_Out), _Fmt, _STD make_format_args(_Args...));
 }
 
 template <output_iterator<const wchar_t&> _OutputIt, class... _Types>
 _OutputIt format_to(_OutputIt _Out, const wstring_view _Fmt, const _Types&... _Args) {
-    _Fmt_iterator_buffer<_OutputIt, wchar_t> _Buf(_STD move(_Out));
-    _STD vformat_to(_Fmt_wit{_Buf}, _Fmt, _STD make_wformat_args(_Args...));
-    return _Buf._Out();
+    return _STD vformat_to(_STD move(_Out), _Fmt, _STD make_wformat_args(_Args...));

```

Comment:
note: this may pull in an old instantiation of vformat_to, due to selectany. However it's OK because that version of format_to will essentially be the first (correct) branch of the if constexpr expression above.

Reply:
selectany?

Reply:
template instantiations are marked with the spiritual equivalent of `__declspec(selectany)` so that when the linker encounters multiple definitions of the symbol (in different object files) it will just pick any of them at random. It actually doesn't apply here, since the implicit conversion happens at the callsight and thus the name mangling is different.

Though if it did it wouldn't be a problem.

Reply:
Well it does apply here, but the name of the selected instantiation changed, so all the selectany symbols under consideration will actually have the same code, even in the old static lib situation.